allennlp/training/trainer.py

```diff
@@ -300,12 +313,16 @@ def _train_epoch(self, epoch: int) -> Dict[str, float]:
         # Set the model to "train" mode.
         self.model.train()
 
-        num_gpus = len(self._cuda_devices)
+        # A `batch_group` has chunks of tensors that form a single batch together for an optimizer
+        # step. A single chunk always contains as many instances as configured in the iterator's
+        # `batch_size` param. The number of chunks in a single `batch_group` is
```

Comment:
This isn't precisely accurate. Iterators create batches that aren't `batch_size` for many reasons. Maybe say something like, for the "simple case of a BasicIterator".

Reply:
"typically contains"?

What are the other possibilities? Gradient accumulation becomes more complicated when the batches aren't all the same size, and this code doesn't handle that case properly. Neither does the multi-GPU code.

Reply:
See this comment: https://github.com/allenai/allennlp/blob/master/training_config/bidirectional_language_model.jsonnet#L28

Reply:
`maximum_samples_per_batch` being the relevant config option.

Reply:
In the case of that LM we also scale our loss internally based on the number of tokens.